lib/auth/grpcserver_test.go

```diff
@@ -1375,6 +1375,120 @@ func TestNodesCRUD(t *testing.T) {
 		// Now expect no nodes to be returned.
 		nodes, err := clt.GetNodes(ctx, apidefaults.Namespace)
 		require.NoError(t, err)
-		require.Equal(t, 0, len(nodes))
+		require.Empty(t, nodes)
+	})
+}
+
+func TestLocksCRUD(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	srv := newTestTLSServer(t)
+
+	clt, err := srv.NewClient(TestAdmin())
+	require.NoError(t, err)
+
+	now := srv.Clock().Now()
+	lock1, err := types.NewLock("lock1", types.LockSpecV2{
+		Target: types.LockTarget{
+			User: "user-A",
+		},
+		Expires: &now,
+	})
+	require.NoError(t, err)
+
+	lock2, err := types.NewLock("lock2", types.LockSpecV2{
+		Target: types.LockTarget{
+			Node: "node",
+		},
+		Message: "node compromised",
+	})
+	require.NoError(t, err)
+
+	t.Run("CreateLock", func(t *testing.T) {
+		// Initially expect no locks to be returned.
+		locks, err := clt.GetLocks(ctx)
+		require.NoError(t, err)
+		require.Empty(t, locks)
+
+		// Create locks.
+		err = clt.UpsertLock(ctx, lock1)
+		require.NoError(t, err)
+
+		err = clt.UpsertLock(ctx, lock2)
+		require.NoError(t, err)
+	})
+
+	// Run LockGetters in nested subtests to allow parallelization.
+	t.Run("LockGetters", func(t *testing.T) {
+		t.Run("GetLocks", func(t *testing.T) {
+			t.Parallel()
+			locks, err := clt.GetLocks(ctx)
+			require.NoError(t, err)
+			require.Len(t, locks, 2)
+			require.Empty(t, cmp.Diff([]types.Lock{lock1, lock2}, locks,
+				cmpopts.IgnoreFields(types.Metadata{}, "ID")))
+		})
+		t.Run("GetLocks with targets", func(t *testing.T) {
+			t.Parallel()
+			// Match both locks with the targets.
+			locks, err := clt.GetLocks(ctx, lock1.Target(), lock2.Target())
+			require.NoError(t, err)
+			require.Len(t, locks, 2)
+			require.Empty(t, cmp.Diff([]types.Lock{lock1, lock2}, locks,
+				cmpopts.IgnoreFields(types.Metadata{}, "ID")))
+
+			// Match only one of the locks.
+			roleTarget := types.LockTarget{Role: "role-A"}
```

Comment:
How is this lock target matching? I don't see "role-A" anywhere above

Reply:
That's correct but the `GetLocks` call below also features `lock1`'s target. The point is that this made up `roleTarget` adds no new results (since none match) but also doesn't prevent those that match (i.e. `lock1`) from being returned.